config.org
#+title: V-Emacs
#+author: Valdigleis (Dk4LL)
#+email: dk4ll@proton.me
#+date: 25/07/2024

* Preparação do ambiente
1. Primeiramente, instale o Emacs no seu ambiente, no Debian isso pode ser feito usando o comando: =sudo apt install emacs-nox emacs-common=   
2. Agora execute os comandos =mkdir ~/.config/emacs= e  =touch ~/.config/emacs/init.el ~/.config/emacs/config.org= para criar os arquivos básicos
3. Cole para o arquivo =~/.config/init= o código:
#+begin_example
(org-babel-load-file
  (expand-file-name "config.org" user-emacs-directory))
#+end_example

*Atenção*: Essa instalação do zsh permite tira o maior proveito desta configuração, a instalação pode ser feita (no Debian) usando o comando =sudo apt install zsh=

* Configurações Básicas
** Desabilitar tela de boas-vindas e ativas o scratch
#+begin_src emacs-lisp
(setq inhibit-startup-message t)
#+end_src

** Habilita o uso de espaços para a tabulação
#+begin_src emacs-lisp
(setq-default indent-tabs-mode nil)
#+end_src

** Habilita deletar texto selecionado ao digitar
#+begin_src emacs-lisp
(delete-selection-mode t)
#+end_src

** Desativar backups e o atuosalve de arquivos globalmente
#+begin_src emacs-lisp
(setq make-backup-files nil)
(setq auto-save-default nil)
#+end_src

** Habilita o salvamento da posição do cursor ao sair
#+begin_src emacs-lisp
(save-place-mode 1)
#+end_src

** Habilita a funcionalidade de lembrar de arquivos recentes
#+begin_src emacs-lisp
(recentf-mode 1)
#+end_src

** Habilita o salvamento do historico de comandos
#+begin_src emacs-lisp
(savehist-mode 1)
#+end_src

** Habilita a atualização automática para os buffers dos arquivos modificados externamente
#+begin_src emacs-lisp
(setopt auto-revert-avoid-polling t)
#+end_src

** Define o intervalo em segundos que o Emacs irá verificar se houve atualização externa dos arquivos
#+begin_src emacs-lisp
(setopt auto-revert-interval 60)
#+end_src

** Habilita o Emacs para verificar informações de controle de versão automaticamente
#+begin_src emacs-lisp
(setopt auto-revert-check-vc-info t)
#+end_src

** Habilita a movimentação entre janelas usando da combinação C-teclas direcionais
#+begin_src emacs-lisp
(windmove-default-keybindings 'control)
#+end_src

** Desabilita a necessidade das sentenças terem que terminar com dois espaços
#+begin_src emacs-lisp
(setopt sentence-end-double-space nil)
#+end_src

** Desabilita o sinal sonoro
#+begin_src emacs-lisp
(setq ring-bell-function 'ignore)
#+end_src

** Desabilita o sinal visual e sonoro do Emacs (Apenas modo gráfico)
#+begin_src emacs-lisp
(when (display-graphic-p)
  (setq visible-bell t))
(when (display-graphic-p)
  (setq visible-bell t))
#+end_src

** Habilita o uso de y e n para confirmações ou negações no Emacs
#+begin_src emacs-lisp
(fset 'yes-or-no-p 'y-or-n-p)
#+end_src

** Habilita o fechamento de pares
#+begin_src emacs-lisp
(electric-pair-mode 1)
#+end_src

** Habilita o uso do ESC para sair do minibuffer e desabilita o C-z
#+begin_src emacs-lisp
(global-set-key (kbd "<escape>") 'keyboard-escape-quit)
(global-unset-key (kbd "C-z"))
#+end_src

* Configuração básica da interface (UI)
** Habilita a exibição do número da coluna
#+begin_src emacs-lisp
(column-number-mode 1)
#+end_src

** Desabilita o menu bar
#+begin_src emacs-lisp
(menu-bar-mode -1)
#+end_src

** Desabilita a tool bar (no modo gráfico)
#+begin_src emacs-lisp
(when (display-graphic-p)
  (tool-bar-mode -1))
#+end_src

** Desabilita a barra de navegação (no modo gráfico)
#+begin_src emacs-lisp
(when (display-graphic-p)
  (scroll-bar-mode -1))
#+end_src

** Desabilita a tooltip (no modo gráfico)
#+begin_src emacs-lisp
(when (display-graphic-p)
  (tooltip-mode -1))
#+end_src

** Desabilita a visualização dos espaços em branco no final das linhas
#+begin_src emacs-lisp
(setopt show-trailing-whitespace nil)
#+end_src

** Melhora efeito visual do texto sublinhado
#+begin_src emacs-lisp
(setopt x-underline-at-descent-line nil)
#+end_src

** Habilita o destaque da linha atual do cursor
#+begin_src emacs-lisp
(global-hl-line-mode t)
#+end_src

** Define a cor usada para o destaque da linha
#+begin_src emacs-lisp
(set-face-background 'hl-line "#5e4a46")
(set-face-foreground 'highlight nil)
#+end_src

** Habilita a quebra visual de linha
#+begin_src emacs-lisp
(global-visual-line-mode t)
#+end_src

** Habilita os indicadores visuais na margem esquerda da janela de edição
#+begin_src emacs-lisp
(setopt indicate-buffer-boundaries 'left)
#+end_src

** Define a borda do Emacs em 10px (no modo gráfico)
#+begin_src emacs-lisp
(when (display-graphic-p)
  (set-fringe-mode 10))
#+end_src

** Configuração para exibição das linhas
#+begin_src emacs-lisp
(global-display-line-numbers-mode 1)
(setq display-line-numbers-type 'relative)
(dolist (mode '(org-mode-hook
                term-mode-hook
                shell-mode-hook
                eshell-mode-hook
                help-mode-hook))
        (add-hook mode (lambda () (display-line-numbers-mode 0))))
#+end_src

** Configuração das fontes usadas (é necessário instalar a FireMono Nerd Font no seus sitema)
#+begin_src emacs-lisp
(set-face-attribute 'default nil :font "FiraMono Nerd Font Mono 13")
(set-face-attribute 'variable-pitch nil :font "FiraMono Nerd Font Mono 13")
(set-face-attribute 'fixed-pitch nil :font "FiraMono Nerd Font Mono 13")
#+end_src

** Transparência da interface
#+begin_src emacs-lisp
(add-to-list 'default-frame-alist '(alpha-background . 90))
#+end_src

** Configura a mensagem no scratch buffer
#+begin_src emacs-lisp
(setq initial-scratch-message ";; ---------------------------------------------\n;; Autor:\n;; Data: \n;; ---------------------------------------------\n")
#+end_src

* Configuração de Backups, custom e fontes
** Define o diretório padrão para backups
#+begin_src emacs-lisp
(setq backup-directory-alist '((".*" . "~/.local/share/Trash/files")))
#+end_src

** Define o arquivo padrão para escrever as configurações
#+begin_src emacs-lisp
(setq custom-file (locate-user-emacs-file "custom.el"))
(load custom-file 'noerror 'nomessage)
#+end_src
    
** Não cria arquivos de bloqueio
#+begin_src emacs-lisp
(setq create-lockfiles nil)
#+end_src

* Configuração para gestão de pacotes
Inicialização (conexão) com os repostiórios de pacotes e realiza a configuração do Helper de gerenciamento "use-package"
#+begin_src emacs-lisp
(require 'package)
(setq package-archives '(("melpa" . "https://melpa.org/packages/")
                          ("elpa" . "https://elpa.gnu.org/packages/")
                          ("gnu" . "https://elpa.gnu.org/packages/")))
(package-initialize)
(unless package-archive-contents
  (package-refresh-contents))
(unless (package-installed-p 'use-package)
   (package-install 'use-package))
(require 'use-package)
(setq use-package-always-ensure t)
#+end_src

* Definir funções elisp usadas no Emacs
Aqui são definidas funções elisp!
** Funcão de recarga as configurações do Emacs
#+begin_src emacs-lisp
(defun dk4ll/reaload-emacs-settings ()
  (interactive)
  (load-file "~/.config/emacs/init.el")) 
#+end_src

** Função para abrir o arquivo config.org
#+begin_src emacs-lisp
(defun dk4ll/open-emacs-config ()
  (interactive)
  (find-file "~/.config/emacs/config.org"))
#+end_src

** Funcão para abrir o diretório de configurações pessoais
#+begin_src emacs-lisp
(defun dk4ll/emacs-personal-folder ()
  (interactive)
  (let ((default-directory "~/.config/emacs/"))
    (call-interactively 'find-file)))
#+end_src

** Função para abrir um terminal no lado esquerdo da janela
#+begin_src emacs-lisp
(defun dk4ll/emacs-right-terminal ()
  (interactive)
  (split-window-right)
  (other-window 1)
  (vterm))
#+end_src

** Função para fechar o terminal e voltar a janela anterior
#+begin_src emacs-lisp
(defun dk4ll/emacs-terminal-close ()
  (interactive)
  (when (string= (buffer-name) "*vterm*")
    (kill-buffer))
  (delete-window)
  (other-window -1))
#+end_src

** Função para aplicar uma rápida organização e formatação no Buffer usando LSP
#+begin_src emacs-lisp
(defun my-lsp-fix-buffer ()
  "Formats buffer and organizes imports using LSP."
  (interactive)
  (lsp-organize-imports)
  (lsp-format-buffer))
#+end_src

* Atalhos para as funções definidas
#+begin_src emacs-lisp
(global-set-key (kbd "C-c t o") 'dk4ll/emacs-right-terminal)
(global-set-key (kbd "C-c t c") 'dk4ll/emacs-terminal-close)
#+end_src

* Instalação dos Pacotes Para funcionalidades
** Vertico
Incrementa o menu de buscas padrão do Emacs.
#+begin_src emacs-lisp
(use-package vertico
  :bind (:map vertico-map
              ("C-i" . vertico-previous)
              ("C-o" . vertico-next)
              ("C-e" . vertico-exit)
         :map minibuffer-local-map
              ("M-h" . backward-kill-word))
  :custom
  (vertico-cycle t)
  :init
  (vertico-mode))
#+end_src

** Marginalia
Adiciona descrição dos itens no meu de buscas padrão do Emacs.
#+begin_src emacs-lisp
(use-package marginalia
  :after vertico
  :custom
  (marginalia-annotators '(marginalia-annotators-heavy marginalia-annotators-light nil))
  :init
  (marginalia-mode))
#+end_src

** Orderless
Adiciona busca fuzzy não ordenada para a busca da busca padrão do Emcas.
#+begin_src emacs-lisp
(use-package orderless
  :config
  (setq completion-styles '(orderless basic)))
#+end_src

** Consult
Adiciona melhoras na experiência de uso na consulta e navegação dentro do Emacs.
#+begin_src emacs-lisp
(use-package consult)
#+end_src

** Which-key
Adiciona a camada de sugestão de comandos no Emacs.
#+begin_src emacs-lisp
(use-package which-key
  :init
  (which-key-mode 1)
  :diminish
  :config
  (setq which-key-side-window-location 'bottom
        which-key-sort-order #'which-key-key-order-alpha
        which-key-allow-imprecise-window-fit nil
        which-key-sort-uppercase-first nil
        which-key-add-column-padding 1
        which-key-max-display-columns nil
        which-key-min-display-lines 8
        which-key-side-window-slot -10
        which-key-side-window-max-height 0.25
        which-key-idle-delay 0.9
        which-key-allow-imprecise-widow-fit nil
        which-key-separator " » "))
#+end_src

** Doom-themes
Adiciona o doom-themes para gerenciar os temas de cores.
#+begin_src emacs-lisp
(use-package doom-themes
  :config
  (setq doom-themes-enable-bold t
        doom-themes-enable-italic t)
  (load-theme 'doom-oceanic-next t)
  (doom-themes-org-config))
#+end_src

** Nano Color Theme
Adiciona o tema de cores nano ao Emacs. Se você optar por usar este tema deverá comentar todas as linhas de código relacionadas ao Doom-themes (acima) e descomentar todas as linhas de código abaixo.
#+begin_src emacs-lisp
;(use-package nano-theme
;  :defer t
;  :quelpa (nano-theme
;           :fetcher github
;           :repo "rougier/nano-theme"))
;(nano-light)
#+end_src

** Doom Modeline
Adiciona a barra de informações melhorada ao Emacs.
#+begin_src emacs-lisp
(use-package doom-modeline
  :hook
  (after-init . doom-modeline-mode)
  :custom
  (doom-modeline-height 35)
  (set-face-attribute 'mode-line nil :font "FiraMono Nerd Font Mono" :height 110)
  (set-face-attribute 'mode-line-inactive nil :font "FiraMono Nerd Font Mono" :height 110)
  (setq doom-modeline-enable-word-count t))
#+end_src

** Nano Modeline
Adiciona uma barra de informações minimalista ao Emacs, se você optar por essa opções descomente o código abaixo e comente todos os códigos relacionados ao pacote Doom Modeline.
#+begin_src emacs-lisp
;(use-package nano-modeline
;  :custom
;  (add-hook 'prog-mode-hook            #'nano-modeline-prog-mode)
;  (add-hook 'text-mode-hook            #'nano-modeline-text-mode)
;  (add-hook 'org-mode-hook             #'nano-modeline-org-mode)
;  (add-hook 'pdf-view-mode-hook        #'nano-modeline-pdf-mode)
;  (add-hook 'mu4e-headers-mode-hook    #'nano-modeline-mu4e-headers-mode)
;  (add-hook 'mu4e-view-mode-hook       #'nano-modeline-mu4e-message-mode)
;  (add-hook 'elfeed-show-mode-hook     #'nano-modeline-elfeed-entry-mode)
;  (add-hook 'elfeed-search-mode-hook   #'nano-modeline-elfeed-search-mode)
;  (add-hook 'term-mode-hook            #'nano-modeline-term-mode)
;  (add-hook 'xwidget-webkit-mode-hook  #'nano-modeline-xwidget-mode)
;  (add-hook 'messages-buffer-mode-hook #'nano-modeline-message-mode)
;  (add-hook 'org-capture-mode-hook     #'nano-modeline-org-capture-mode)
;  (add-hook 'org-agenda-mode-hook      #'nano-modeline-org-agenda-mode))
;(nano-modeline-text-mode t)
#+end_src

** Diminish
Melhora exibição na doom-modeline.
#+begin_src emacs-lisp
(use-package diminish)
#+end_src

** Rainbow Delimiters
Ativa multiplas cores nos delimitadores de código aninhados.
#+begin_src emacs-lisp
(use-package rainbow-delimiters
  :hook ((prog-mode . rainbow-delimiters-mode)
         (emcas-lisp-mode . rainbow-delimiters-mode)
         (cloujure-mode . raindow-delimiters-mode)))
#+end_src

** Company
Disponibiliza recursos avançados para completar.
#+begin_src emacs-lisp
(use-package company
  :defer 2
  :diminish
  :custom
  (company-begin-commands '(self-insert-command))
  (company-idle-delay .1)
  (company-minimum-prefix-length 2)
  (company-show-numbers t)
  (company-tooltip-align-annotations 't)
  (global-company-mode t))

(use-package company-box
  :after company
  :diminish
  :hook (company-mode . company-box-mode))
#+end_src

** Treemacs
Disponibiliza ferramenta lateral do tipo explore files na visão hierárquica de árvore.
#+begin_src emacs-lisp
  (use-package treemacs-nerd-icons
    :config
    (treemacs-load-theme "nerd-icons"))

  (use-package treemacs
    :bind
    (:map global-map
          ("C-c f" . treemacs))
    :config
    (setq treemacs-no-png-images t
          treemacs-is-never-other-window nil
          treemacs-show-hidden-files t
          treemacs-git-mode 'deferred))

  (require 'treemacs-nerd-icons)
  (treemacs-load-theme "nerd-icons")

#+end_src

** Vterm
Habilita um terminal interno no Emacs, porém, em sistemas baseados em Debian (e talvez outros Unix-like) são necessários as seguintes depedências:
1. [[https://cmake.org/][CMAKE]].
2. [[https://www.gnu.org/software/libtool/][libtool]].
3. [[https://www.leonerd.org.uk/code/libvterm/][libvterm]], cuidado pois existe uma versão para [[https://github.com/akermu/emacs-libvterm?tab=readme-ov-file][Emacs]] e uma para [[https://github.com/neovim/libvterm?tab=readme-ov-file][NeoVim]].
Para usuários de [[https://www.debian.org/][Debian]] (ou derivados) todas as depedências podem ser obtidas via gereciador de pacotes apt. Após instalar as depedências lembre de modificar no código abaixo para usar seu interpretador de conmandos favorito.

#+begin_src emacs-lisp
(use-package vterm  
  :config
  (setq shell-file-name "/usr/bin/zsh"
        vterm-max-scrollback 5000))
#+end_src

** Flycheck
Para a verificação de sintaxe, linters e etc. Trabalha junto com os servidores de linguagens.
#+begin_src emacs-lisp
(use-package flycheck
  :hook (prog-mode-hook . flycheck-mode))
#+end_src

** Projectile
Para o gerenciamento de projetos, isto é, gerencia o caminho dos projetos!
#+begin_src emacs-lisp
(use-package projectile
  :diminish projectile-mode
  :config
  (projectile-mode))
#+end_src

** Magit
Para gerenciamento e uso do [[https://git-scm.com/][Git]] embarcado no Emacs.
#+begin_src emacs-lisp
(use-package magit)
#+end_src

** Rainbow Mode
#+begin_src emacs-lisp
(use-package rainbow-mode
:diminish
:hook org-mode prog-mode)
#+end_src

** Undo-tree
Pacote para exibir visualmente uma árvore de edição contendo todos os estados assumido pelo buffer, incluido os estados que sofreram undo.
#+begin_src emacs-lisp
(use-package undo-tree
  :config (global-undo-tree-mode t))
#+end_src

* Configuração do Orgmode
** Ativar o orgmode
#+begin_src emacs-lisp
(require 'org)
#+end_src

** Adicionar extensão .org ao Org Mode
#+begin_src emacs-lisp
(add-to-list 'auto-mode-alist '("\\.org\\'" . org-mode))
#+end_src

** Habilitar que os ambientes sejam iniciado com fold fechado
#+begin_src emacs-lisp
(setq org-startup-folded t)
#+end_src

** Habilitar folding para códigos dentro de blocos de código
#+begin_src emacs-lisp
(setq org-src-fontify-natively t)
(setq org-src-tab-acts-natively t)
(setq org-edit-src-content-indentation 0)
#+end_src

** Desabilita o fechamento automático do pares de delimitadores
#+begin_src emacs-lisp
(add-hook 'org-mode-hook (lambda ()
                           (setq-local electric-pair-inhibit-predicate
                                       '(lambda (c)
                                          (if (char-equal c ?<) t (,electric-pair-inhibit-predicate c))))))
#+end_src  

** Desabilita a indentação automática e zera as margens
#+begin_src emacs-lisp
(electric-indent-mode -1)
(setq org-edit-src-content-indentation 0)
#+end_src

** Habilita o org-tempo
#+begin_src emacs-lisp
(require 'org-tempo)
#+end_src

** Instala, Habilita e Configura o org-bullets
#+begin_src emacs-lisp
(add-hook 'org-mode-hook 'org-indent-mode)
(use-package org-bullets
  :custom
  (org-bullets-bullet-list '("" "" "◆" "◇" "▪" "▪" "▪")))
(add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))
#+end_src

* Suporte de Linguagens
** Markdown 
Para utilização do markdown-mode do Emacs é necessário que você possua instalado em seu sistema o [[https://pandoc.org/][pandoc]], e o mesmo deve estar liberado para uso com o comando =pandoc=, se você usa algum outro alias para o pandoc no seu sistema, então você deverá modificar o código abaixo.
#+begin_src emacs-lisp
(use-package markdown-mode
  :mode "\\.md\\'"
  :config
  (setq markdown-command "pandoc")
  (setq markdown-asymmetric-header t)
  (setq markdown-header-scaling t)
  (setq markdown-enable-math t))
#+end_src

Ativa a disponibilidade para a função de pré-visualização dos arquivos markdown
#+begin_src emacs-lisp
(use-package markdown-preview-mode
  :commands markdown-preview)
#+end_src

** HTML/CSS
#+begin_src emacs-lisp
(use-package web-mode
  :ensure t
  :mode (("\\.html?\\'" . web-mode)
         ("\\.phtml\\'" . web-mode)
         ("\\.php\\'" . web-mode)
         ("\\.css\\'" . web-mode))
  :config
  (setq web-mode-markup-indent-offset 2)
  (setq web-mode-css-indent-offset 2)
  (setq web-mode-code-indent-offset 2)
  (setq web-mode-enable-css-colorization t)
  (setq web-mode-enable-auto-pairing t)
  (setq web-mode-enable-comment-keywords t)
  (setq web-mode-enable-current-element-highlight t)
  (setq web-mode-comment-style 2)
  (setq web-mode-style-padding 1)
  (setq web-mode-script-padding 1)
  (setq web-mode-block-padding 0))
#+end_src

** JSON
#+begin_src emacs-lisp
(use-package json-mode)
#+end_src

** Ocaml
Para gerenciar todas as depedências do ambiente de Ocaml recomendo a instalação previa do [[https://opam.ocaml.org/][opam]] em seu sistema, de forma que o mesmo esteja disponível via PATH.
#+begin_src emacs-lisp
(use-package tuareg
  :custom
  (tuareg-opam-insinuate t)
  :config)
(use-package dune-format)
(use-package reason-mode)
#+end_src

** Latex
Habilita uma utilização suave e totalmente integral ao LaTeX.
#+begin_src emacs-lisp
(use-package auctex
  :defer t)
#+end_src

* Flycheck
** Verificação de sintaxe com flycheck
#+begin_src emacs-lisp
(use-package flycheck
  :init (global-flycheck-mode))
#+end_src

* LSP
** Python
É necessário que o [[https://github.com/pappasam/jedi-language-server][Jedi-LSP]] esteja instalado no seu sistema, e disponível via PATH do sistema.
#+begin_src emacs-lisp
(use-package lsp-jedi)
(setq python-indent-offset 2)
#+end_src

** Haskell
É necessário que o GHCi e o [[https://github.com/haskell/haskell-language-server/][Haskell language server]] estejam instalados no seu sistema, e disponíveis via PATH do sistema, recomendo que seja feita a instalação via [[https://www.haskell.org/ghcup/][ghcup]].
#+begin_src emacs-lisp
(use-package lsp-haskell)
#+end_src

** Java

config.el
(setq inhibit-startup-message t)

(setq-default indent-tabs-mode nil)

(delete-selection-mode t)

(setq make-backup-files nil)
(setq auto-save-default nil)

(save-place-mode 1)

(recentf-mode 1)

(savehist-mode 1)

(setopt auto-revert-avoid-polling t)

(setopt auto-revert-interval 60)

(setopt auto-revert-check-vc-info t)

(windmove-default-keybindings 'control)

(setopt sentence-end-double-space nil)

(setq ring-bell-function 'ignore)

(when (display-graphic-p)
  (setq visible-bell t))
(when (display-graphic-p)
  (setq visible-bell t))

(fset 'yes-or-no-p 'y-or-n-p)

(electric-pair-mode 1)

(global-set-key (kbd "<escape>") 'keyboard-escape-quit)
(global-unset-key (kbd "C-z"))

(column-number-mode 1)

(menu-bar-mode -1)

(when (display-graphic-p)
  (tool-bar-mode -1))

(when (display-graphic-p)
  (scroll-bar-mode -1))

(when (display-graphic-p)
  (tooltip-mode -1))

(setopt show-trailing-whitespace nil)

(setopt x-underline-at-descent-line nil)

(global-hl-line-mode t)

(set-face-background 'hl-line "#5e4a46")
(set-face-foreground 'highlight nil)

(global-visual-line-mode t)

(setopt indicate-buffer-boundaries 'left)

(when (display-graphic-p)
  (set-fringe-mode 10))

(global-display-line-numbers-mode 1)
(setq display-line-numbers-type 'relative)
(dolist (mode '(org-mode-hook
                term-mode-hook
                shell-mode-hook
                eshell-mode-hook
                help-mode-hook))
        (add-hook mode (lambda () (display-line-numbers-mode 0))))

(set-face-attribute 'default nil :font "FiraMono Nerd Font Mono 13")
(set-face-attribute 'variable-pitch nil :font "FiraMono Nerd Font Mono 13")
(set-face-attribute 'fixed-pitch nil :font "FiraMono Nerd Font Mono 13")

(add-to-list 'default-frame-alist '(alpha-background . 90))

(setq initial-scratch-message ";; ---------------------------------------------\n;; Autor:\n;; Data: \n;; ---------------------------------------------\n")

(setq backup-directory-alist '((".*" . "~/.local/share/Trash/files")))

(setq custom-file (locate-user-emacs-file "custom.el"))
(load custom-file 'noerror 'nomessage)

(setq create-lockfiles nil)

(require 'package)
(setq package-archives '(("melpa" . "https://melpa.org/packages/")
                          ("elpa" . "https://elpa.gnu.org/packages/")
                          ("gnu" . "https://elpa.gnu.org/packages/")))
(package-initialize)
(unless package-archive-contents
  (package-refresh-contents))
(unless (package-installed-p 'use-package)
   (package-install 'use-package))
(require 'use-package)
(setq use-package-always-ensure t)

(defun dk4ll/reaload-emacs-settings ()
  (interactive)
  (load-file "~/.config/emacs/init.el"))

(defun dk4ll/open-emacs-config ()
  (interactive)
  (find-file "~/.config/emacs/config.org"))

(defun dk4ll/emacs-personal-folder ()
  (interactive)
  (let ((default-directory "~/.config/emacs/"))
    (call-interactively 'find-file)))

(defun dk4ll/emacs-right-terminal ()
  (interactive)
  (split-window-right)
  (other-window 1)
  (vterm))

(defun dk4ll/emacs-terminal-close ()
  (interactive)
  (when (string= (buffer-name) "*vterm*")
    (kill-buffer))
  (delete-window)
  (other-window -1))

(defun my-lsp-fix-buffer ()
  "Formats buffer and organizes imports using LSP."
  (interactive)
  (lsp-organize-imports)
  (lsp-format-buffer))

(global-set-key (kbd "C-c t o") 'dk4ll/emacs-right-terminal)
(global-set-key (kbd "C-c t c") 'dk4ll/emacs-terminal-close)

(use-package vertico
  :bind (:map vertico-map
              ("C-i" . vertico-previous)
              ("C-o" . vertico-next)
              ("C-e" . vertico-exit)
         :map minibuffer-local-map
              ("M-h" . backward-kill-word))
  :custom
  (vertico-cycle t)
  :init
  (vertico-mode))

(use-package marginalia
  :after vertico
  :custom
  (marginalia-annotators '(marginalia-annotators-heavy marginalia-annotators-light nil))
  :init
  (marginalia-mode))

(use-package orderless
  :config
  (setq completion-styles '(orderless basic)))

(use-package consult)

(use-package which-key
  :init
  (which-key-mode 1)
  :diminish
  :config
  (setq which-key-side-window-location 'bottom
        which-key-sort-order #'which-key-key-order-alpha
        which-key-allow-imprecise-window-fit nil
        which-key-sort-uppercase-first nil
        which-key-add-column-padding 1
        which-key-max-display-columns nil
        which-key-min-display-lines 8
        which-key-side-window-slot -10
        which-key-side-window-max-height 0.25
        which-key-idle-delay 0.9
        which-key-allow-imprecise-widow-fit nil
        which-key-separator " » "))

(use-package doom-themes
  :config
  (setq doom-themes-enable-bold t
        doom-themes-enable-italic t)
  (load-theme 'doom-oceanic-next t)
  (doom-themes-org-config))

;(use-package nano-theme
;  :defer t
;  :quelpa (nano-theme
;           :fetcher github
;           :repo "rougier/nano-theme"))
;(nano-light)

(use-package doom-modeline
  :hook
  (after-init . doom-modeline-mode)
  :custom
  (doom-modeline-height 35)
  (set-face-attribute 'mode-line nil :font "FiraMono Nerd Font Mono" :height 110)
  (set-face-attribute 'mode-line-inactive nil :font "FiraMono Nerd Font Mono" :height 110)
  (setq doom-modeline-enable-word-count t))

;(use-package nano-modeline
;  :custom
;  (add-hook 'prog-mode-hook            #'nano-modeline-prog-mode)
;  (add-hook 'text-mode-hook            #'nano-modeline-text-mode)
;  (add-hook 'org-mode-hook             #'nano-modeline-org-mode)
;  (add-hook 'pdf-view-mode-hook        #'nano-modeline-pdf-mode)
;  (add-hook 'mu4e-headers-mode-hook    #'nano-modeline-mu4e-headers-mode)
;  (add-hook 'mu4e-view-mode-hook       #'nano-modeline-mu4e-message-mode)
;  (add-hook 'elfeed-show-mode-hook     #'nano-modeline-elfeed-entry-mode)
;  (add-hook 'elfeed-search-mode-hook   #'nano-modeline-elfeed-search-mode)
;  (add-hook 'term-mode-hook            #'nano-modeline-term-mode)
;  (add-hook 'xwidget-webkit-mode-hook  #'nano-modeline-xwidget-mode)
;  (add-hook 'messages-buffer-mode-hook #'nano-modeline-message-mode)
;  (add-hook 'org-capture-mode-hook     #'nano-modeline-org-capture-mode)
;  (add-hook 'org-agenda-mode-hook      #'nano-modeline-org-agenda-mode))
;(nano-modeline-text-mode t)

(use-package diminish)

(use-package rainbow-delimiters
  :hook ((prog-mode . rainbow-delimiters-mode)
         (emcas-lisp-mode . rainbow-delimiters-mode)
         (cloujure-mode . raindow-delimiters-mode)))

(use-package company
  :defer 2
  :diminish
  :custom
  (company-begin-commands '(self-insert-command))
  (company-idle-delay .1)
  (company-minimum-prefix-length 2)
  (company-show-numbers t)
  (company-tooltip-align-annotations 't)
  (global-company-mode t))

(use-package company-box
  :after company
  :diminish
  :hook (company-mode . company-box-mode))

(use-package treemacs-nerd-icons
  :config
  (treemacs-load-theme "nerd-icons"))

(use-package treemacs
  :bind
  (:map global-map
        ("C-c f" . treemacs))
  :config
  (setq treemacs-no-png-images t
        treemacs-is-never-other-window nil
        treemacs-show-hidden-files t
        treemacs-git-mode 'deferred))

(require 'treemacs-nerd-icons)
(treemacs-load-theme "nerd-icons")

(use-package vterm  
  :config
  (setq shell-file-name "/usr/bin/zsh"
        vterm-max-scrollback 5000))

(use-package flycheck
  :hook (prog-mode-hook . flycheck-mode))

(use-package projectile
  :diminish projectile-mode
  :config
  (projectile-mode))

(use-package magit)

(use-package rainbow-mode
:diminish
:hook org-mode prog-mode)

(use-package undo-tree
  :config (global-undo-tree-mode t))

(require 'org)

(add-to-list 'auto-mode-alist '("\\.org\\'" . org-mode))

(setq org-startup-folded t)

(setq org-src-fontify-natively t)
(setq org-src-tab-acts-natively t)
(setq org-edit-src-content-indentation 0)

(add-hook 'org-mode-hook (lambda ()
                           (setq-local electric-pair-inhibit-predicate
                                       '(lambda (c)
                                          (if (char-equal c ?<) t (,electric-pair-inhibit-predicate c))))))

(electric-indent-mode -1)
(setq org-edit-src-content-indentation 0)

(require 'org-tempo)

(add-hook 'org-mode-hook 'org-indent-mode)
(use-package org-bullets
  :custom
  (org-bullets-bullet-list '("" "" "◆" "◇" "▪" "▪" "▪")))
(add-hook 'org-mode-hook (lambda () (org-bullets-mode 1)))

(use-package markdown-mode
  :mode "\\.md\\'"
  :config
  (setq markdown-command "pandoc")
  (setq markdown-asymmetric-header t)
  (setq markdown-header-scaling t)
  (setq markdown-enable-math t))

(use-package markdown-preview-mode
  :commands markdown-preview)

(use-package web-mode
  :ensure t
  :mode (("\\.html?\\'" . web-mode)
         ("\\.phtml\\'" . web-mode)
         ("\\.php\\'" . web-mode)
         ("\\.css\\'" . web-mode))
  :config
  (setq web-mode-markup-indent-offset 2)
  (setq web-mode-css-indent-offset 2)
  (setq web-mode-code-indent-offset 2)
  (setq web-mode-enable-css-colorization t)
  (setq web-mode-enable-auto-pairing t)
  (setq web-mode-enable-comment-keywords t)
  (setq web-mode-enable-current-element-highlight t)
  (setq web-mode-comment-style 2)
  (setq web-mode-style-padding 1)
  (setq web-mode-script-padding 1)
  (setq web-mode-block-padding 0))

(use-package json-mode)

(use-package tuareg
  :custom
  (tuareg-opam-insinuate t)
  :config)
(use-package dune-format)
(use-package reason-mode)

(use-package auctex
  :defer t)

(use-package flycheck
  :init (global-flycheck-mode))

(use-package lsp-jedi)
(setq python-indent-offset 2)

(use-package lsp-haskell)

(use-package lsp-java)

(use-package lsp-latex)
(setq tex-indent-basic 2)

(use-package php-mode)
(with-eval-after-load 'php-mode
  (custom-set-variables '(lsp-phpactor-path "~/.apps/phpactor/"))
  (add-hook 'php-mode-hook #'init-php-mode))

(use-package lsp-mode
  :after flycheck
  :commands lsp
  :bind (("C-c l ," . flycheck-previous-error)
         ("C-c l ." . flycheck-next-error)
         ("C-c l d" . lsp-find-definition)
         ("C-c l r" . lsp-find-references)
         ("C-c l p" . lsp-describe-thing-at-point)
         ("C-c l i" . lsp-find-implementation)
         ("C-c l R" . lsp-rename)
         ("C-c l f" . my-lsp-fix-buffer))
  :hook ((c-mode . lsp)
	     (c++-mode . lsp)
         (python-mode . lsp)
         (haskell-mode . lsp)
         (web-mode . lsp)
         (java-mode . lsp)
         (tex-mode . lsp)
         (bibtex-mode . lsp)
         (tuareg-mode . lsp)
         (caml-mode . lsp)
         (php-mode . lsp)
         (reason-mode . lsp)
         (before-save . lsp-organize-imports))
  :config
  (setq lsp-clients-clangd-executable "/usr/bin/clangd")
  (setq lsp-completion-provider :capf))

(use-package lsp-ui
  :commands lsp-ui-mode
  :after lsp-mode
  :hook (lsp-mode . lsp-ui-mode)
  :config
  (lsp-ui-doc-enable t))
